pkg/controller/garbagecollector/reconciler.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package garbagecollector
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/gardener/gardener-resource-manager/pkg/controller/garbagecollector/references"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type reconciler struct {
+	log          logr.Logger
+	syncPeriod   time.Duration
+	targetClient client.Client
+}
+
+func (r *reconciler) InjectLogger(l logr.Logger) error {
+	r.log = l.WithName(ControllerName)
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	var (
+		labels                  = client.MatchingLabels{references.LabelKeyGarbageCollectable: references.LabelValueGarbageCollectable}
+		objectsToGarbageCollect = map[string]struct{}{}
```

Comment:
Why aren't you using a proper struct type as the key instead? Would save you the `objectId` and `detailsFromObjectId` funcs and fiddling with structured strings.

Reply:
I can do that, but then I would end up with a `[]object` type requiring to change the logic in https://github.com/gardener/gardener-resource-manager/pull/127/files#diff-52d6d64c1f4a2a5b1c1de5a74016a526b6e279cbbd70621852c512dbed2bc105R77-R94. Today, it can easily delete the candidate from the `objectsToGarbageCollect` map because the key (=object id) is unique. Using `[]object` would complicate the entire logic, wouldn't it?